File: emm/web_ui/src/api/testing/queueItems.ts
/**
 * Обёртки над публичным API очереди `testing_service` (`/queue-items/*` и
 * управление очередью стенда на `/test-stands/{id}/resume-queue`).
 *
 * Источник истины — `testing_service/src/api/v1/endpoints/queue_items.py`
 * и `test_stands.py`.
 */
import { apiGet, apiPatch, apiPost } from "@/api/client";
import type { ActiveQueueMode, QueueVerdict, TestingPaginatedResponse, VerdictSource } from "./types";
export { getCurrentQueueItem, findActiveQueueItemForServer } from "@/api/testing/testStands";

/**
 * Запрошенное прерывание исполняющегося теста. Проставляется `skip`/`pause`,
 * когда item реально исполняется на стенде — воркер подтвердит прерывание
 * асинхронно, до этого item остаётся в `running`.
 */
export type QueueInterruptAction = "skip" | "pause";

export interface PublicQueueItem {
  log_status?: "available" | "rotated" | "pending" | "missing";
  test_code?: string | null; test_name?: string | null; is_current?: boolean;
  id: string; test_id: string; stand_id: string; test_run_id: string | null;
  retry_of_id: string | null; debug_mode: boolean; prepare_only?: boolean; state: string;
  interrupt_action?: QueueInterruptAction | null;
  rc: string | null; kernel: string | null; mode: string | null;
  created_at: string; started_at: string | null; finished_at: string | null; error: string | null;
  /** Место в очереди стенда (меньше — раньше); значимо между не терминальными элементами. */
  position?: number;
  /**
   * Вердикт: `passed`/`failed`/`unknown`. `unknown` — «результат не
   * определён»: у запуска нет прогона в Zephyr (debug), исход смотреть в
   * логе/Confluence. `null` — ещё не вынесен.
   */
  verdict?: QueueVerdict | null;
  /** Откуда взят исход: статус в Zephyr или код выхода `starter.sh`. */
  verdict_source?: VerdictSource | null;
  /** Статус тест-кейса в Zephyr как есть (последний прочитанный). */
  zephyr_status_raw?: string | null;
  verdict_resolved_at?: string | null;
  /** Многоступенчатый тест: индекс текущего шага с 0 и число шагов. */
  current_step_index?: number;
  step_count?: number | null;
}
export interface QueueLaunchRequest {
  request_id: string; test_id: string; stand_id: string; os_version_id: string;
  kernel: string; debug_mode: boolean;
  /**
   * Легаси testenv-режим: только откатить/подготовить стенд, не запускать
   * тест. Вместо результата воркер оставляет на стенде `command.txt` с
   * командой, которая была бы запущена.
   */
  prepare_only?: boolean;
  /**
   * Забрать стенд у текущего держателя (`busy`/`testing_done`). Без роли
   * department_admin/`admin` testing_service своего отдела сервер отклоняет
   * запрос отдельным `FORCE_LAUNCH_DENIED`; `updating` и чужой `acs` не
   * отбираются никогда (`STAND_TAKEOVER_NOT_ALLOWED`).
   */
  force?: boolean;
  /**
   * Режим при уже активной очереди стенда: `append` — в конец, `replace` —
   * очистить очередь, прервать текущий тест и начать сразу. Не задан — админ
   * получает 409 `STAND_QUEUE_ACTIVE`, остальные встают в конец.
   */
  on_active_queue?: ActiveQueueMode;
}
export interface QueueItemsQuery {
  kind?: "standalone" | "campaign" | "all";
  os_version_id?: string; os_version_name?: string; kernel?: string;
  test_run_id?: string; stand_id?: string; test_id?: string;
  attempt_id?: string; retry_of_id?: string;
  created_from?: string; created_until?: string;
  states?: string[]; debug_mode?: boolean; q?: string;
  order?: "asc" | "desc"; limit?: number; offset?: number;
}
export function listQueueItems(query: QueueItemsQuery = {}) {
  const params = new URLSearchParams({ kind: "standalone", limit: "50" });
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === "") continue;
    if (Array.isArray(value)) value.forEach((item) => params.append(key, item));
    else params.set(key, String(value));
  }
  return apiGet<TestingPaginatedResponse<PublicQueueItem>>(`/testing/v1/queue-items?${params}`);
}
export function launchQueueItem(body: QueueLaunchRequest) {
  return apiPost<PublicQueueItem>("/testing/v1/queue-items", body);
}
export function retryQueueItem(id: string, requestId: string) {
  return apiPost<PublicQueueItem>(`/testing/v1/queue-items/${encodeURIComponent(id)}/retry`, { request_id: requestId });
}

/**
 * `POST /queue-items/{id}/skip` — пропустить тест. Если он уже исполняется на
 * стенде, ответ приходит с прежним `state: "running"` и
 * `interrupt_action: "skip"` — прерывание подтвердит воркер, перечитывать
 * состояние нужно опросом. Из `queued`/`preparing`/`ready` item уходит в
 * `skipped` сразу, стенд продолжает со следующего.
 */
export function skipQueueItem(id: string) {
  return apiPost<PublicQueueItem>(`/testing/v1/queue-items/${encodeURIComponent(id)}/skip`, {});
}

/**
 * `POST /queue-items/{id}/pause` — остановить тест без исхода. Семантика
 * ответа та же, что у `skip`, целевое состояние — `paused`; стенд на
 * следующий item не переходит, пока не вызван `resumeStandQueue`.
 */
export function pauseQueueItem(id: string) {
  return apiPost<PublicQueueItem>(`/testing/v1/queue-items/${encodeURIComponent(id)}/pause`, {});
}

/**
 * `POST /test-stands/{id}/resume-queue` — снять стенд с паузы: единственный
 * `paused`-item переставляется в конец очереди этого стенда и возвращается в
 * `queued`. 409 `STAND_NOT_PAUSED`, если у стенда нет остановленного item'а.
 */
export function resumeStandQueue(standId: string) {
  return apiPost<PublicQueueItem>(`/testing/v1/test-stands/${encodeURIComponent(standId)}/resume-queue`, {});
}

/**
 * `PATCH /test-stands/{id}/queue/order` — новый порядок ещё не начатых
 * (`queued`) элементов стенда (D15). `queueItemIds` — все текущие `queued`
 * стенда в желаемом порядке; активный/остановленный элемент не передаётся и
 * не переставляется (409 `QUEUE_ITEM_NOT_QUEUED`). Если очередь успела
 * измениться (головной стартовал, в хвост встал новый) — 409
 * `QUEUE_ORDER_STALE`, очередь нужно перечитать.
 */
export function reorderStandQueue(standId: string, queueItemIds: string[]) {
  return apiPatch<{ items: PublicQueueItem[] }>(
    `/testing/v1/test-stands/${encodeURIComponent(standId)}/queue/order`,
    { queue_item_ids: queueItemIds },
  );
}

File: emm/testing_service/src/api/v1/endpoints/queue_items.py
"""Публичная очередь, отдельная от протокола testing_worker."""

from typing import Literal
from pydantic import AwareDatetime
from src.core.exceptions import DomainValidationError
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from src.dependencies.auth import CurrentUserIdentity
from src.dependencies.db import get_db
from src.schemas.common import OkResponse, PaginatedResponse
from src.schemas.public_queue import (
    PublicQueueItem,
    QueueLaunchRequest,
    QueueRetryRequest,
)
from src.repositories import queue_item as repo
from src.services import public_queue as svc, log_availability

router = APIRouter(prefix="/queue-items")


@router.post("", response_model=PublicQueueItem, status_code=201)
async def launch(
    body: QueueLaunchRequest,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
):
    item = await svc.launch(db, identity, body)
    logs = await log_availability.for_items(db, [item])
    return svc.response(item).model_copy(update={"log_status": logs[item.id]})


@router.post("/{item_id}/retry", response_model=PublicQueueItem, status_code=201)
async def retry(
    item_id: str,
    body: QueueRetryRequest,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
):
    item = await svc.retry(db, identity, item_id, body)
    logs = await log_availability.for_items(db, [item])
    return svc.response(item).model_copy(update={"log_status": logs[item.id]})


@router.post(
    "/{item_id}/skip",
    response_model=PublicQueueItem,
    summary="Пропустить элемент очереди",
    description=(
        "Снимает тест с исполнения без исхода: элемент уходит в `skipped`, "
        "стенд сразу продолжает со следующего. Если элемент прямо сейчас "
        "исполняется на стенде (`running`), моментально оборвать его нельзя — "
        "ответ вернёт тот же `running`-элемент с `interrupt_action=\"skip\"`, "
        "а фактический обрыв сделает `testing_worker` при ближайшем опросе."
    ),
    responses={
        403: {"description": "Нет прав на стенд этого элемента."},
        404: {"description": "Элемент очереди не найден."},
        409: {"description": "QUEUE_ITEM_NOT_ACTIVE — элемент уже завершён."},
    },
)
async def skip(
    item_id: str,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
):
    item = await svc.skip(db, identity, item_id)
    logs = await log_availability.for_items(db, [item])
    return svc.response(item).model_copy(update={"log_status": logs[item.id]})


@router.post(
    "/{item_id}/pause",
    response_model=PublicQueueItem,
    summary="Остановить элемент очереди",
    description=(
        "То же прерывание, что и `skip`, но исход не фиксируется: элемент "
        "уходит в `paused` и остаётся в очереди стенда, стенд встаёт и ждёт "
        "`POST /test-stands/{stand_id}/resume-queue`. Для `running`-элемента "
        "ответ несёт `interrupt_action=\"pause\"` — см. `skip`."
    ),
    responses={
        403: {"description": "Нет прав на стенд этого элемента."},
        404: {"description": "Элемент очереди не найден."},
        409: {"description": "QUEUE_ITEM_NOT_ACTIVE — элемент уже завершён."},
    },
)
async def pause(
    item_id: str,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
):
    item = await svc.pause(db, identity, item_id)
    logs = await log_availability.for_items(db, [item])
    return svc.response(item).model_copy(update={"log_status": logs[item.id]})


@router.delete(
    "/{item_id}",
    response_model=OkResponse,
    summary="Удалить элемент очереди",
    description=(
        "Убирает элемент из очереди насовсем — в отличие от `skip`, "
        "терминальной записи не остаётся. Работает для ещё не стартовавших "
        "и уже терминальных элементов; для того, что прямо сейчас занимает "
        "стенд (`preparing`/`ready`/`running`), сначала нужен `skip`/`pause`."
    ),
    responses={
        403: {"description": "Нет прав на стенд этого элемента."},
        404: {"description": "Элемент очереди не найден."},
        409: {"description": "QUEUE_ITEM_IN_PROGRESS — элемент сейчас занимает стенд."},
    },
)
async def delete_queue_item(
    item_id: str,
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
):
    await svc.delete(db, identity, item_id)
    return OkResponse()


@router.get("", response_model=PaginatedResponse[PublicQueueItem])
async def list_items(
    identity: CurrentUserIdentity,
    db: AsyncSession = Depends(get_db),
    kind: Literal["standalone", "campaign", "all"] = "standalone",
    test_run_id: str | None = None,
    os_version_id: str | None = None,
    os_version_name: str | None = Query(None, max_length=64),
    kernel: str | None = None,
    stand_id: str | None = None,
    test_id: str | None = None,
    attempt_id: str | None = None,
    retry_of_id: str | None = None,
    created_from: AwareDatetime | None = None,
    created_until: AwareDatetime | None = None,
    states: list[Literal[
        "queued", "preparing", "ready", "running",
        "succeeded", "failed", "skipped", "paused",
    ]] | None = Query(None),
    debug_mode: bool | None = None,
    q: str | None = Query(None, max_length=200),
    order: Literal["asc", "desc"] = "desc",
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    if created_from and created_until and created_from >= created_until:
        raise DomainValidationError(
            error_code="INVALID_TIME_RANGE", message="Начало периода должно быть раньше конца"
        )
    items, total = await repo.list_for_department(
        db,
        identity.department_id or "",
        kind=kind,
        test_run_id=test_run_id,
        limit=limit,
        offset=offset,
        os_version_id=os_version_id, os_version_name=os_version_name, kernel=kernel, stand_id=stand_id, test_id=test_id, attempt_id=attempt_id,
        retry_of_id=retry_of_id, created_from=created_from, created_until=created_until,
        states=states, debug_mode=debug_mode, q=q, order=order,
    )
    logs = await log_availability.for_items(db, [row[0] for row in items])
    return PaginatedResponse(
        items=[svc.response(item).model_copy(update={
            "test_code": code, "test_name": name, "is_current": current, "log_status": logs[item.id],
        }) for item, code, name, current in items],
        total=total,
        limit=limit,
        offset=offset,
    )
